flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/sinks/SelectTableSinkBase.java

```diff
@@ -28,61 +28,102 @@
 
 	private final TableSchema tableSchema;
-	private final CollectSinkOperatorFactory<Row> factory;
-	private final CollectResultIterator<Row> iterator;
+	protected final DataFormatConverters.DataFormatConverter<RowData, Row> converter;
+
+	private final CollectSinkOperatorFactory<T> factory;
+	private final CollectResultIterator<T> iterator;
 
 	@SuppressWarnings("unchecked")
-	public SelectTableSinkBase(TableSchema tableSchema) {
+	public SelectTableSinkBase(TableSchema schema, TypeSerializer<T> typeSerializer) {
 		this.tableSchema = SelectTableSinkSchemaConverter.convertTimeAttributeToRegularTimestamp(
-			SelectTableSinkSchemaConverter.changeDefaultConversionClass(tableSchema));
+				SelectTableSinkSchemaConverter.changeDefaultConversionClass(schema));
+		this.converter = DataFormatConverters.getConverterForDataType(this.tableSchema.toPhysicalRowDataType());
 
-		TypeSerializer<Row> typeSerializer = (TypeSerializer<Row>) TypeInfoDataTypeConverter
-			.fromDataTypeToTypeInfo(this.tableSchema.toRowDataType())
-			.createSerializer(new ExecutionConfig());
 		String accumulatorName = "tableResultCollect_" + UUID.randomUUID();
-
 		this.factory = new CollectSinkOperatorFactory<>(typeSerializer, accumulatorName);
 		CollectSinkOperator<Row> operator = (CollectSinkOperator<Row>) factory.getOperator();
 		this.iterator = new CollectResultIterator<>(operator.getOperatorIdFuture(), typeSerializer, accumulatorName);
 	}
 
 	@Override
-	public DataType getConsumedDataType() {
-		return tableSchema.toRowDataType();
+	public TableSchema getTableSchema() {
+		return tableSchema;
 	}
 
 	@Override
-	public TableSchema getTableSchema() {
-		return tableSchema;
+	public TableSink<T> configure(String[] fieldNames, TypeInformation<?>[] fieldTypes) {
+		throw new UnsupportedOperationException();
 	}
 
-	protected DataStreamSink<?> consumeDataStream(DataStream<Row> dataStream) {
-		CollectStreamSink<Row> sink = new CollectStreamSink<>(dataStream, factory);
+	@Override
+	public DataStreamSink<?> consumeDataStream(DataStream<T> dataStream) {
+		CollectStreamSink<?> sink = new CollectStreamSink<>(dataStream, factory);
 		dataStream.getExecutionEnvironment().addOperator(sink.getTransformation());
 		return sink.name("Select table sink");
 	}
 
-	@Override
-	public void setJobClient(JobClient jobClient) {
-		iterator.setJobClient(jobClient);
+	public SelectResultProvider getSelectResultProvider() {
+		return new SelectResultProvider() {
+			@Override
+			public void setJobClient(JobClient jobClient) {
+				iterator.setJobClient(jobClient);
+			}
+
+			@Override
+			public Iterator<Row> getResultIterator() {
+				return new RowIteratorWrapper(iterator);
+			}
+		};
 	}
 
-	@Override
-	public Iterator<Row> getResultIterator() {
-		return iterator;
+	/**
+	 * An Iterator wrapper class that converts Iterator&lt;T&gt; to Iterator&lt;Row&gt;.
+	 */
+	private class RowIteratorWrapper implements Iterator<Row> {
```

Comment:
Should also implement `AutoClosable`, otherwise job related resources will not be disposed if user reads only a partial of results.